org.osgi.test.common/src/main/java/org/osgi/test/common/event/CloneUtil.java

```diff
@@ -0,0 +1,114 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ *******************************************************************************/
+
+package org.osgi.test.common.event;
+
+import java.lang.reflect.InvocationHandler;
+import java.lang.reflect.Method;
+import java.lang.reflect.Proxy;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.stream.Stream;
+
+import org.osgi.framework.Bundle;
+import org.osgi.framework.BundleEvent;
+import org.osgi.framework.FrameworkEvent;
+import org.osgi.framework.ServiceEvent;
+import org.osgi.framework.ServiceReference;
+
+class CloneUtil {
+	public static ServiceEvent clone(ServiceEvent serviceEvent) {
+		return new ServiceEvent(serviceEvent.getType(), clone(serviceEvent.getServiceReference()));
+	}
+
+	public static BundleEvent clone(BundleEvent bundleEvent) {
+		return new BundleEvent(bundleEvent.getType(), clone(bundleEvent.getBundle()), clone(bundleEvent.getOrigin()));
+	}
+
+	public static FrameworkEvent clone(FrameworkEvent frameworkEvent) {
+		return new FrameworkEvent(frameworkEvent.getType(), clone(frameworkEvent.getBundle()),
+			frameworkEvent.getThrowable());
+	}
+
+	public static ServiceReference<?> clone(ServiceReference<?> serviceReference) {
+		Map<String, Object> props = new HashMap<>();
+		if (serviceReference.getPropertyKeys() != null) {
+			for (String key : serviceReference.getPropertyKeys()) {
+				props.put(key, serviceReference.getProperty(key));
+			}
+		}
+		Bundle bundle = clone(serviceReference.getBundle());
+		Bundle[] usingBundles = serviceReference.getUsingBundles() == null ? null
+			: Stream.of(serviceReference.getUsingBundles())
+				.map(b -> clone(b))
+				.toArray(Bundle[]::new);
+		ServiceReference<?> serviceReferenceProxyInstance = (ServiceReference<?>) Proxy
+			.newProxyInstance(ServiceReference.class.getClassLoader(), new Class[] {
+				ServiceReference.class
+			}, new InvocationHandler() {
+				@Override
+				public Object invoke(Object proxy, Method method, Object[] args) throws Throwable {
+					if (method.getName()
+						.equals("getProperty")) {
+						return props.get(args[0]);
+					}
+					if (method.getName()
+						.equals("getPropertyKeys")) {
+						return props.keySet()
+							.stream()
+							.toArray(String[]::new);
+					}
+					if (method.getName()
+						.equals("getBundle")) {
+						return bundle;
+					}
+					if (method.getName()
+						.equals("getUsingBundles")) {
+						return usingBundles;
+					}
+					return method.invoke(proxy, args);
```

Comment:
Isn't this an infinite recursion? You call the same method again on the same proxy object?

Reply:
SURE! It must use the origin ServiceReference.